transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueEventLoop.java

```diff
@@ -92,16 +92,28 @@ public int get() throws Exception {
 
     void add(AbstractKQueueChannel ch) {
         assert inEventLoop();
-        channels.put(ch.fd().intValue(), ch);
+        AbstractKQueueChannel old = channels.put(ch.fd().intValue(), ch);
+        // We either expect to have no Channel in the map with the same FD or that the FD of the old Channel is already
+        // closed.
+        assert old == null || !old.fd().isOpen();
     }
 
     void evSet(AbstractKQueueChannel ch, short filter, short flags, int fflags) {
-        changeList.evSet(ch, filter, flags, fflags);
+        // Only try to add to changeList if the FD is still open, if not we already closed it in the meantime.
+        if (ch.fd().isOpen()) {
+            changeList.evSet(ch, filter, flags, fflags);
+        }
     }
 
     void remove(AbstractKQueueChannel ch) {
         assert inEventLoop();
-        channels.remove(ch.fd().intValue());
+        int fd = ch.fd().intValue();
+        // Remove only if that's the same channel
+        // Due to file descriptor reuse another channel might have overwritten
+        // the removed channel in the map
+        if (channels.get(fd) == ch) {
```

Comment:
I would like to keep this as consistent as possible with `EpollEventLoop`, and if it is not possible clarify why in the comments. There was a bunch of investigation related to life time of FDs in kevent related to when things will be cleaned up. We may be able to avoid some work in general here (e.g. always doing a get before remove here, always checking open in `evSet`), and I would like to increase confidence we are not masking a sequencing issue elsewhere. The investigation was done in the context of the experimental JNI reference approach but I think it is worth doing the same investigation now that we are retaining `Channel` objects in a map in this class.

Reply:
@Scottmitch PTAL again... I compared both implementations and while doing so I also noticed a bug in the EpollEventLoop implementation which could suffer from FD re-use as well. Beside this it also did not correctly remove old entries from the map sometimes which could lead to not have these GCed. 

Now the implementations should behave the same. I also ported the test to run for both impls.